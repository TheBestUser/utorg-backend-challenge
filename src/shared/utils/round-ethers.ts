import { parseFixed } from '@ethersproject/bignumber';

export function ceilEthers(value: string, decimals = 18): string {
  const [integer, decimal] = parseFloatParts(value);

  const toCeilDigit = decimal[decimals] ?? '0';

  if (toCeilDigit === '0') {
    const decimalFixed = decimal.slice(0, decimals).padEnd(decimals, '0');
    return formatFloatParts(integer, decimalFixed, decimals);
  }

  const concatedAmount =
    integer + decimal.slice(0, decimals + 1).padEnd(decimals + 1, '0');
  const ceilString = parseFixed(concatedAmount, 0)
    .add(9)
    .toString()
    .slice(0, -1);

  const [integerFixed, decimalFixed] = [
    ceilString.substring(0, ceilString.length - decimals),
    ceilString.substring(ceilString.length - decimals),
  ];

  return formatFloatParts(integerFixed, decimalFixed, decimals);
}

export function floorEthers(value: string, decimals = 18): string {
  const [integer, decimal] = parseFloatParts(value);
  const decimalFixed = decimal.slice(0, decimals).padEnd(decimals, '0');
  return formatFloatParts(integer, decimalFixed, decimals);
}

function parseFloatParts(value: string): [string, string] {
  const [integer, decimal] = value.split('.');

  return [integer || '0', decimal || '0'];
}
function formatFloatParts(
  integer: string,
  decimal: string,
  decimals: number,
): string {
  return decimals <= 0 ? integer : `${integer}.${decimal}`;
}
