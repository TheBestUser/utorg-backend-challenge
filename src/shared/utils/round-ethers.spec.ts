import { ceilEthers, floorEthers } from './round-ethers';

describe('round-ethers', () => {
  test.each`
    value               | decimals
    ${'999.999999999'}  | ${0}
    ${'999.999999999'}  | ${6}
    ${'1000.999999999'} | ${6}
    ${'1000.1234567'}   | ${6}
    ${'1000.1234560'}   | ${6}
    ${'1000.0'}         | ${6}
    ${'1000'}           | ${6}
    ${'0'}              | ${6}
    ${'0.0'}            | ${6}
    ${''}               | ${6}
  `('value: "$value", decimals: $decimals', ({ value, decimals }) => {
    expect(ceilEthers(value, decimals)).toMatchSnapshot('ceilEthers');
    expect(floorEthers(value, decimals)).toMatchSnapshot('floorEthers');
  });
});
