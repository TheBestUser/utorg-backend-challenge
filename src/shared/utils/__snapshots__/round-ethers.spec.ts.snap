// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`round-ethers value: "", decimals: 6: ceilEthers 1`] = `"0.000000"`;

exports[`round-ethers value: "", decimals: 6: floorEthers 1`] = `"0.000000"`;

exports[`round-ethers value: "0", decimals: 6: ceilEthers 1`] = `"0.000000"`;

exports[`round-ethers value: "0", decimals: 6: floorEthers 1`] = `"0.000000"`;

exports[`round-ethers value: "0.0", decimals: 6: ceilEthers 1`] = `"0.000000"`;

exports[`round-ethers value: "0.0", decimals: 6: floorEthers 1`] = `"0.000000"`;

exports[`round-ethers value: "999.999999999", decimals: 0: ceilEthers 1`] = `"1000"`;

exports[`round-ethers value: "999.999999999", decimals: 0: floorEthers 1`] = `"999"`;

exports[`round-ethers value: "999.999999999", decimals: 6: ceilEthers 1`] = `"1000.000000"`;

exports[`round-ethers value: "999.999999999", decimals: 6: floorEthers 1`] = `"999.999999"`;

exports[`round-ethers value: "1000", decimals: 6: ceilEthers 1`] = `"1000.000000"`;

exports[`round-ethers value: "1000", decimals: 6: floorEthers 1`] = `"1000.000000"`;

exports[`round-ethers value: "1000.0", decimals: 6: ceilEthers 1`] = `"1000.000000"`;

exports[`round-ethers value: "1000.0", decimals: 6: floorEthers 1`] = `"1000.000000"`;

exports[`round-ethers value: "1000.1234560", decimals: 6: ceilEthers 1`] = `"1000.123456"`;

exports[`round-ethers value: "1000.1234560", decimals: 6: floorEthers 1`] = `"1000.123456"`;

exports[`round-ethers value: "1000.1234567", decimals: 6: ceilEthers 1`] = `"1000.123457"`;

exports[`round-ethers value: "1000.1234567", decimals: 6: floorEthers 1`] = `"1000.123456"`;

exports[`round-ethers value: "1000.999999999", decimals: 6: ceilEthers 1`] = `"1001.000000"`;

exports[`round-ethers value: "1000.999999999", decimals: 6: floorEthers 1`] = `"1000.999999"`;
